fxrunner/src/lib/shutdown/windows.rs

```diff
@@ -0,0 +1,240 @@
+    InitiateSystemShutdown,
+}
+
+#[derive(Debug, Display)]
+#[display(fmt = "{}: {}", kind, source)]
+pub struct ShutdownError {
+    kind: ShutdownErrorKind,
+    source: WindowsError,
+}
+
+impl Error for ShutdownError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        Some(&self.source)
+    }
+}
+
+/// A wrapper around a HANDLE that automatically closes.
+struct Handle(HANDLE);
+
+impl Handle {
+    /// Create a new null handle.
+    pub fn null() -> Self {
+        Handle(null_mut())
+    }
+
+    /// Return the underlying `HANDLE`.
+    pub fn as_ptr(&self) -> HANDLE {
+        self.0
+    }
+
+    /// Return a mutable pointer to the underlying `HANDLE`, allowing it to be
+    /// used as an output parameter.
+    pub fn as_out_ptr(&mut self) -> *mut HANDLE {
+        &mut self.0 as *mut HANDLE
+    }
+}
+
+impl Drop for Handle {
+    fn drop(&mut self) {
+        if !self.0.is_null() {
+            let rv = unsafe { CloseHandle(self.0 as HANDLE) };
+            assert!(rv != 0);
+
+            self.0 = null_mut();
+        }
+    }
+}
+
+// See: https://docs.microsoft.com/en-us/windows/win32/shutdown/how-to-shut-down-the-system
+pub(super) fn initiate_restart(reason: &str) -> Result<(), ShutdownError> {
+    let mut token = Handle::null();
+    let mut privs = unsafe { std::mem::zeroed::<TOKEN_PRIVILEGES>() };
+
+    let name = CString::new(SE_SHUTDOWN_NAME).unwrap();
+    let success = unsafe {
+        OpenProcessToken(
+            GetCurrentProcess(),
+            TOKEN_ADJUST_PRIVILEGES | TOKEN_QUERY,
+            token.as_out_ptr(),
+        ) != 0
+    };
+    if !success {
+        return Err(ShutdownError {
+            kind: ShutdownErrorKind::OpenProcessToken,
+            source: get_last_error(),
+        });
+    }
+
+    let success = unsafe {
+        // We cast `name`'s underlying pointer to `LPSTR` (`*mut c_char`) here,
+        // but it is safe because it will not be modified through the pointer.
+        LookupPrivilegeValueA(
+            null_mut(),
+            name.as_ptr(),
+            &mut privs.Privileges[0].Luid as *mut LUID,
+        ) != 0
+    };
+    if !success {
+        return Err(ShutdownError {
+            kind: ShutdownErrorKind::LookupPrivilegeValue,
+            source: get_last_error(),
+        });
+    }
+
+    privs.PrivilegeCount = 1;
+    privs.Privileges[0].Attributes = SE_PRIVILEGE_ENABLED;
+
+    let success = unsafe {
+        AdjustTokenPrivileges(
+            token.as_ptr(),
+            false as BOOL,
+            &mut privs as *mut TOKEN_PRIVILEGES,
+            0 as DWORD,
+            null_mut::<TOKEN_PRIVILEGES>(),
+            null_mut::<DWORD>(),
+        ) != 0
+    };
+
+    if !success {
+        return Err(ShutdownError {
+            kind: ShutdownErrorKind::AdjustTokenPrivileges,
+            source: get_last_error(),
+        });
+    }
+
+    let reason_ptr = CString::new(reason).unwrap().into_raw();
```

Comment:
So I originally wrote this last fall and my memory is a little hazy. I think because `into_raw()` gives a `*mut c_char` and `as_ptr()` gives a `*const c_char`. The API takes an `LPSTR` (`*mut c_char`) and so used `into_raw()`.

However, the API doesn't mutate the underlying string and it therefore it should be safe to use `as_ptr()` and cast it to a `*mut c_char`.